packages/frontend/app/(dashboard)/course/[cid]/insights/components/charts/GanttChartComponent.tsx

```diff
@@ -0,0 +1,140 @@
+import {
+  AxisChartClasses,
+  ChartDataType,
+  GanttChartComponentProps,
+} from '@/app/(dashboard)/course/[cid]/insights/utils/types'
+import React, { useMemo } from 'react'
+import {
+  generateAxisRange,
+  generateTickRange,
+} from '@/app/(dashboard)/course/[cid]/insights/utils/functions'
+import {
+  ChartContainer,
+  ChartTooltip,
+  ChartTooltipContent,
+} from '@/app/components/ui/chart'
+import {
+  CartesianGrid,
+  Scatter,
+  ScatterChart,
+  XAxis,
+  YAxis,
+  ZAxis,
+} from 'recharts'
+
+const GanttChartComponent: React.FC<GanttChartComponentProps> = ({ props }) => {
+  const {
+    yKey,
+    zKey,
+    chartConfig,
+    chartData,
+    size,
+    numCategories,
+    aspectRatio,
+  } = props
+
+  let {
+    includeTooltip,
+    labelFormatter,
+    xTickFormatter,
+    yTickFormatter,
+    valueFormatter,
+    xRange,
+  } = props
+
+  includeTooltip ??= true
+  labelFormatter ??= (label: string) => label
+  xTickFormatter ??= (tick: string) => tick
+  yTickFormatter ??= (tick: string) => tick
+  valueFormatter ??= (value: string) => value
+  xRange ??= (chartData: ChartDataType[]) =>
+    generateAxisRange(chartData, ['key'], 10)
```

Comment:
huh makes sense that you can assign variables to functions with it too. Also ngl i kinda laughed when i saw it from how absurd this looks compared to the code i'm used to seeing, but then again i think i've had several of these moments working with react over the past some months, and this also makes a lot of sense

Reply:
you laugh? you laugh while there's nullish coalescing operators used to assign functions to variables?

Reply:
_statements dreamed up by the utterly deranged_